service/src/initialized_service.rs

```diff
@@ -47,8 +51,48 @@ pub async fn start_is_initialized_server(port: u16) -> ServiceResult<()> {
 	Ok(())
 }
 
-/// Set initialized handler value to true.
-pub fn set_initialized() {
-	let mut initialized_lock = INITIALIZED_HANDLE.write();
-	*initialized_lock = true;
+/// Trait to query of a worker is considered fully initialized.
+pub trait IsInitialized {
+	fn is_initialized(&self) -> bool;
+}
+
+/// Tracker for initialization. Used by components that ensure these steps were taken.
+pub trait TrackInitialization {
+	fn registered_on_parentchain(&self);
+
+	fn sidechain_block_produced(&self);
+
+	fn worker_for_shard_registered(&self);
+}
+
+#[derive(Default)]
+pub struct InitializationHandler {
+	registered_on_parentchain: Mutex<bool>,
+	sidechain_block_produced: Mutex<bool>,
+	worker_for_shard_registered: Mutex<bool>,
```

Comment:
What prevents us from using `RwLock`?

Reply:
You're absolutely right, I misinterpreted an error message I had. I had `RwLock` at first (which is my default guard), but an error message threw me off, and I thought I had to use `Mutex` - but I was wrong.

In the past I ran into a case where this mattered ([source](https://stackoverflow.com/questions/50704279/when-or-why-should-i-use-a-mutex-over-an-rwlock)):
* Mutex [requires T: Send](https://doc.rust-lang.org/std/sync/struct.Mutex.html#implementations) to be Sync,
* RwLock [requires T to be Send and Sync](https://doc.rust-lang.org/std/sync/struct.RwLock.html#implementations) to be itself Sync.

But here we have `bool` as our `T`, so it implements `Send` and `Sync` and we can use `RwLock` 😅

Reply:
Ah, thans for explaining this implementation detail. I learned something new! :D